### PR TITLE
Kill SQLite

### PR DESCRIPTION
**Pull request by seanh**
Use Dockerised PostgreSQL in development and testing and on Travis CI. Also use Dockerized Elasticsearch and NSQ in development.

### CODE REVIEW

**Comment by seanh** [comment]
Running `make test` repeatedly is fine but there's a problem with this when trying to run just some model tests:

    py.test h/accounts/test/models_test.py

gets

```
    @pytest.fixture()
    def db_session(request, settings):
        """SQLAlchemy session."""
        engine = db.make_engine(settings)
>       db.bind_engine(engine, should_create=True, should_drop=True)

h/conftest.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
h/db.py:67: in bind_engine
    base.metadata.drop_all(engine)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/schema.py:3714: in drop_all
    tables=tables)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:1856: in _run_visitor
    conn._run_visitor(visitorcallable, element, **kwargs)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:1481: in _run_visitor
    **kwargs).traverse_single(element)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/visitors.py:121: in traverse_single
    return meth(obj, **kw)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py:871: in visit_metadata
    table, drop_ok=True, _is_metadata_operation=True)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/visitors.py:121: in traverse_single
    return meth(obj, **kw)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py:918: in visit_table
    self.connection.execute(DropTable(table))
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:914: in execute
    return meth(self, multiparams, params)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py:68: in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:968: in _execute_ddl
    compiled
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:1146: in _execute_context
    context)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:1341: in _handle_dbapi_exception
    exc_info
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py:199: in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:1139: in _execute_context
    context)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2 object at 0x7f5597438a10>, cursor = <cursor object at 0x7f5597470a00; closed: -1>, statement = '\nDROP TABLE "user"', parameters = {}
context = <sqlalchemy.dialects.postgresql.psycopg2.PGExecutionContext_psycopg2 object at 0x7f5597e1e390>

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       InternalError: (psycopg2.InternalError) cannot drop table "user" because other objects depend on it
E       DETAIL:  constraint group_creator_id_fkey on table "group" depends on table "user"
E       constraint user_group_user_id_fkey on table user_group depends on table "user"
E       HINT:  Use DROP ... CASCADE to drop the dependent objects too.
E        [SQL: '\nDROP TABLE "user"']

../../../.virtualenvs/h/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py:450: InternalError
```

**Comment by seanh** [comment]
I've fixed the tests issue, this is ready to go

File: .travis.yml [hunk 1]
```diff
@@ -10,6 +12,8 @@ install:
   - pip install coveralls pyramid_redis_sessions
   - npm install
   - make
+before_script:
+  - psql -c 'create database htest;' -U postgres
```
Comment (nickstenning) [inline .travis.yml]: `createdb htest` would probably be clearer.

File: h/db.py [hunk 2]
```diff
@@ -62,6 +63,8 @@ def bind_engine(engine,
     session.configure(bind=engine)
     base.metadata.bind = engine
     if should_drop:
+        transaction.commit()
```
Comment (nickstenning) [inline h/db.py]: What's the need to commit the transaction here?

File: setup.py [hunk 3]
```diff
@@ -70,7 +70,7 @@ def run_tests(self):
 ]
 
 DEV_EXTRAS = ['pyramid_debugtoolbar>=2.1', 'prospector[with_pyroma]', 'pep257',
-              'pyramid_multiauth', 'sphinxcontrib-httpdomain']
+              'pyramid_multiauth', 'sphinxcontrib-httpdomain', 'psycopg2']
```
Comment (nickstenning) [inline setup.py]: Given this is now in `DEV_EXTRAS` and in `requirements.txt`, I think we should probably move `psycopg2` and `psycogreen` both into setup.py as regular dependencies.

File: h/db.py (hunk 2 above)
Comment (seanh) [inline h/db.py]: That's there to close any db connections, without it our model tests just hang forever

Comment (seanh) [inline h/db.py]: Ok, I can move this into the `db_session` fixture instead

**Comment by seanh** [comment]
@nickstenning Ok, all done

**Comment by nickstenning** [comment]
Ok, this branch needs rebasing, so if you could remove the finalizer from `db_session` and simply put the:

```python
transaction.commit()
db.Session.remove()
```

at the top of the fixture function, then this is good to go.

**Comment by seanh** [comment]
Done

File: docs/hacking/install.rst [hunk 4]
```diff
@@ -116,43 +117,85 @@ Install compass:
...
+      .. code-block:: bash
+
+         docker start postgres elasticsearch nsqd
+
+3. Create the `htest` database in the ``postgres`` container. This is needed
+   to run the h tests:
+
+   .. code-block:: bash
+
+      docker run -it --link postgres:postgres --rm postgres sh -c 'exec psql -h "$POSTGRES_PORT_5432_TCP_ADDR" -p "$POSTGRES_PORT_5432_TCP_PORT" -U postgres -c "CREATE DATABASE htest;"'
```
Comment (tilgovi) [inline docs/hacking/install.rst]: Shouldn't the tests create and drop this themselves?

**Comment by tilgovi** [comment]
I don't see any change to `h/config.py`. Shouldn't it be made to pick up the linked postgres container?

**Pull request merged**